In [1]:
import h2o
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321 ..... not found.
Attempting to start a local H2O server...
; Java HotSpot(TM) 64-Bit Server VM (build 25.211-b12, mixed mode)
  Starting server from C:\Users\maria\Anaconda3\lib\site-packages\h2o\backend\bin\h2o.jar
  Ice root: C:\Users\maria\AppData\Local\Temp\tmp_6llexzy
  JVM stdout: C:\Users\maria\AppData\Local\Temp\tmp_6llexzy\h2o_maria_started_from_python.out
  JVM stderr: C:\Users\maria\AppData\Local\Temp\tmp_6llexzy\h2o_maria_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  ------------------------------------------
H2O cluster uptime:         17 secs
H2O cluster timezone:       Africa/Luanda
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.2
H2O cluster version age:    28 days, 9 hours and 32 minutes
H2O cluster name:           H2O_from_python_maria_96h9yl
H2O cluster total nodes:    1
H2O cluster free memory:    892 Mb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         locked, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, Algos, AutoML, Core V3, Core V4
Python version:             3.6.5 final
--------------------------  ------------------------------------------

# Problem Description

Based on some available features, we will like to determine if some people will click on our AD when using the internet.

In [3]:
data = h2o.import_file("advertising.csv")
data.head(5)# The default head() command displays the first 10 rows.

Parse progress: |█████████████████████████████████████████████████████████| 100%


Daily Time Spent on Site Age Area Income Daily Internet Usage Ad Topic Line City Male Country Timestamp Clicked on Ad 
 
 
 68.95 35 61833.9 256.09 Cloned 5thgeneration orchestration Wrightburgh 0 Tunisia 2016-03-27 00:53:11 0 
 80.23 31 68441.9 193.77 Monitored national standardization West Jodi 1 Nauru 2016-04-04 01:39:02 0 
 69.47 26 59785.9 236.5 Organic bottom-line service-desk Davidton 0 San Marino 2016-03-13 20:35:42 0 
 74.15 29 54806.2 245.89 Triple-buffered reciprocal time-frame West Terrifurt 1 Italy 2016-01-10 02:31:19 0 
 68.37 35 73890 225.58 Robust logistical utilization South Manuel 0 Iceland 2016-06-03 03:36:18 0

# Important features about H2o.AI 
(http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/glm.html)

1) Standardization: The Standardization of numerical values is enabled by default. If you do not use standardization, the results can include components that are dominated by variables that appear to have larger variances relative to other attributes as a matter of scale, rather than true contribution. Only advanced users should disable this option.

2) Missing values handling: In H2O, the Deep Learning and GLM algorithms will either skip or mean-impute rows with NA values. This option defaults to MeanImputation. But you can select the skip option.

3) Unwanted Columns: The ignored_columns parameter is used to specify an array of column names that should be ignored.

4) Collinearity : Collinear columns can cause problems during model fitting. The preferred way to deal with collinearity (and the default H2O behavior) is to add regularization. However, if you want a non-regularized solution, you can choose to automatically remove collinear columns by enabling the "remove_collinear_columns" option. This option can only be used when solver=IRLSM and with no regularization (lambda=0).

In [5]:
# check the number of missing values
#Depending on the selected missing value handling policy, 
#they are either imputed mean or the whole row is skipped. The default behavior is Mean Imputation. 
print('missing:', data.isna().sum())

missing: 0.0


In [6]:
data.describe()

Rows:1000
Cols:10




Daily Time Spent on Site Age Area Income Daily Internet Usage Ad Topic Line City Male Country Timestamp Clicked on Ad 
 
 
 type real int real real string string int enum time int 
 mins 32.6 19.0 13996.5 104.78 NaN NaN 0.0 1451616730000.0 0.0 
 mean 65.0002 36.009 55000.00008 180.00009999999997 NaN NaN 0.481 1460284446636.0 0.5 
 maxs 91.43 61.0 79484.8 269.96 NaN NaN 1.0 1469319736000.0 1.0 
 sigma 15.853614567500209 8.78556231012592 13414.634022282358 43.9023393019801 NaN NaN 0.49988887654046565 5089978400.060859 0.5002501876563868 
 zeros 0 0 0 0 0 0 519 0 500 
 missing 0 0 0 0 0 0 0 0 0 0 
 0 68.95 35.0 61833.9 256.09 Cloned 5thgeneration orchestration Wrightburgh 0.0 Tunisia 2016-03-27 00:53:11 0.0 
 1 80.23 31.0 68441.85 193.77 Monitored national standardization West Jodi 1.0 Nauru 2016-04-04 01:39:02 0.0 
 2 69.47 26.0 59785.94 236.5 Organic bottom-line service-desk Davidton 0.0 San Marino 2016-03-13 20:35:42 0.0 
 3 74.15 29.0 54806.18 245.89 Triple-buffered reciprocal time-frame West Terrifurt 1.0 Italy 2016-01-10 02:31:19 0.0 
 4 68.37 35.0 73889.99 225.58 Robust logistical utilization South Manuel 0.0 Iceland 2016-06-03 03:36:18 0.0 
 5 59.99 23.0 59761.56 226.74 Sharable client-driven software Jamieberg 1.0 Norway 2016-05-19 14:30:17 0.0 
 6 88.91 33.0 53852.85 208.36 Enhanced dedicated support Brandonstad 0.0 Myanmar 2016-01-28 20:59:32 0.0 
 7 66.0 48.0 24593.33 131.76 Reactive local challenge Port Jefferybury 1.0 Australia 2016-03-07 01:40:15 1.0 
 8 74.53 30.0 68862.0 221.51 Configurable coherent function West Colin 1.0 Grenada 2016-04-18 09:33:42 0.0 
 9 69.88 20.0 55642.32 183.82 Mandatory homogeneous architecture Ramirezton 1.0 Ghana 2016-07-11 01:42:51 0.0

In [4]:
i_split = data.split_frame(ratios = [0.8], seed = 1234)
train = i_split[0] # using 80% for training
test = i_split[1] #rest 20% for testing
print(train.shape, test.shape)

(802, 10) (198, 10)


In [5]:
#set the response column to to levels
data['Clicked on Ad'] = data['Clicked on Ad'].asfactor()

# set the predictor names and the response column name
predictors= ['Daily Time Spent on Site', 'Age', 'Area Income','Daily Internet Usage', 'Male']
response = 'Clicked on Ad'

# split into train and validation sets
#train, valid= airlines.split_frame(ratios = [.8], seed = 1234)
i_split = data.split_frame(ratios = [0.8], seed = 1234)
train = i_split[0] # using 80% for training
test = i_split[1] #rest 20% for testing


In [18]:
from h2o.estimators.glm import H2OGeneralizedLinearEstimator

glm_model = H2OGeneralizedLinearEstimator(family= "binomial", compute_p_values = True, lambda_ = 0)
glm_model.train(predictors, response, training_frame= train, validation_frame = test)
print(glm_model)


glm Model Build progress: |███████████████████████████████████████████████| 100%
Model Details
H2OGeneralizedLinearEstimator :  Generalized Linear Modeling
Model Key:  GLM_model_python_1557913803799_7


ModelMetricsBinomialGLM: glm
** Reported on train data. **

MSE: 0.023463935384969305
RMSE: 0.15317942219818334
LogLoss: 0.09085402217815138
Null degrees of freedom: 801
Residual degrees of freedom: 796
Null deviance: 1111.8080776181512
Residual deviance: 145.72985157375481
AIC: 157.72985157375481
AUC: 0.9920429599318412
pr_auc: 0.8045255494902589
Gini: 0.9840859198636824
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.5386312169517615: 


0    1    Error    Rate
-----  ---  ---  -------  ------------
0      394  7    0.0175   (7.0/401.0)
1      14   387  0.0349   (14.0/401.0)
Total  408  394  0.0262   (21.0/802.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.538631     0.973585  135
max f2                       0.18102      0.97284   162
max f0point5                 0.674908     0.982555  128
max accuracy                 0.538631     0.973815  135
max precision                0.999992     1         0
max recall                   0.0068782    1         328
max specificity              0.999992     1         0
max absolute_mcc             0.538631     0.947775  135
max min_per_class_accuracy   0.394074     0.967581  142
max mean_per_class_accuracy  0.538631     0.973815  135

Gains/Lift Table: Avg response rate: 50.00 %, avg score: 50.00 %



group    cumulative_data_fraction    lower_threshold    lift    cumulative_lift    response_rate    score       cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain    cumulative_gain
--  -------  --------------------------  -----------------  ------  -----------------  ---------------  ----------  --------------------------  ------------------  --------------  -------------------------  ------  -----------------
    1        0.0112219                   0.999996           2       2                  1                0.999997    1                           0.999997            0.0224439       0.0224439                  100     100
    2        0.021197                    0.999996           2       2                  1                0.999996    1                           0.999997            0.0199501       0.042394                   100     100
    3        0.0311721                   0.999994           2       2                  1                0.999995    1                           0.999996            0.0199501       0.0623441                  100     100
    4        0.0411471                   0.999994           2       2                  1                0.999994    1                           0.999996            0.0199501       0.0822943                  100     100
    5        0.0511222                   0.999993           2       2                  1                0.999993    1                           0.999995            0.0199501       0.102244                   100     100
    6        0.100998                    0.999982           2       2                  1                0.999988    1                           0.999992            0.0997506       0.201995                   100     100
    7        0.150873                    0.999957           2       2                  1                0.99997     1                           0.999985            0.0997506       0.301746                   100     100
    8        0.200748                    0.999908           2       2                  1                0.999933    1                           0.999972            0.0997506       0.401496                   100     100
    9        0.300499                    0.999363           2       2                  1                0.999737    1                           0.999894            0.199501        0.600998                   100     100
    10       0.400249                    0.989756           2       2                  1                0.99716     1                           0.999213            0.199501        0.800499                   100     100
    11       0.5                         0.380415           1.675   1.93516            0.8375           0.839592    0.967581                    0.967368            0.167082        0.967581                   67.5    93.5162
    12       0.599751                    0.0353507          0.2     1.64657            0.1              0.125808    0.823285                    0.8274              0.0199501       0.987531                   -80     64.657
    13       0.699501                    0.0124569          0.05    1.41889            0.025            0.0197527   0.709447                    0.712227            0.00498753      0.992519                   -95     41.8895
    14       0.799252                    0.00681061         0.075   1.25117            0.0375           0.00938268  0.625585                    0.624509            0.0074813       1                          -92.5   25.117
    15       0.899002                    0.00406632         0       1.11234            0                0.00549892  0.556172                    0.555825            0               1                          -100    11.2344
    16       1                           0.001754           0       1                  0                0.00308702  0.5                         0.5                 0               1                          -100    0



ModelMetricsBinomialGLM: glm
** Reported on validation data. **

MSE: 0.02369804444066122
RMSE: 0.1539416916909166
LogLoss: 0.09510143545409801
Null degrees of freedom: 197
Residual degrees of freedom: 192
Null deviance: 274.48628350173834
Residual deviance: 37.66016843982282
AIC: 49.66016843982282
AUC: 0.9907152331394756
pr_auc: 0.9832395618791827
Gini: 0.9814304662789513
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.20978118730958129: 


0    1    Error    Rate
-----  ---  ---  -------  -----------
0      97   2    0.0202   (2.0/99.0)
1      3    96   0.0303   (3.0/99.0)
Total  100  98   0.0253   (5.0/198.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.209781     0.974619  97
max f2                       0.209781     0.97166   97
max f0point5                 0.688854     0.989474  93
max accuracy                 0.688854     0.974747  93
max precision                0.999998     1         0
max recall                   0.00588656   1         158
max specificity              0.999998     1         0
max absolute_mcc             0.688854     0.950708  93
max min_per_class_accuracy   0.209781     0.969697  97
max mean_per_class_accuracy  0.688854     0.974747  93

Gains/Lift Table: Avg response rate: 50.00 %, avg score: 48.82 %



group    cumulative_data_fraction    lower_threshold    lift    cumulative_lift    response_rate    score       cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain    cumulative_gain
--  -------  --------------------------  -----------------  ------  -----------------  ---------------  ----------  --------------------------  ------------------  --------------  -------------------------  ------  -----------------
    1        0.010101                    0.999998           2       2                  1                0.999998    1                           0.999998            0.020202        0.020202                   100     100
    2        0.020202                    0.999996           2       2                  1                0.999997    1                           0.999998            0.020202        0.040404                   100     100
    3        0.030303                    0.999994           2       2                  1                0.999995    1                           0.999997            0.020202        0.0606061                  100     100
    4        0.040404                    0.999992           2       2                  1                0.999993    1                           0.999996            0.020202        0.0808081                  100     100
    5        0.0505051                   0.999988           2       2                  1                0.999991    1                           0.999995            0.020202        0.10101                    100     100
    6        0.10101                     0.999969           2       2                  1                0.999978    1                           0.999987            0.10101         0.20202                    100     100
    7        0.151515                    0.999959           2       2                  1                0.999965    1                           0.999979            0.10101         0.30303                    100     100
    8        0.20202                     0.999928           2       2                  1                0.999943    1                           0.99997             0.10101         0.40404                    100     100
    9        0.30303                     0.999633           2       2                  1                0.99981     1                           0.999917            0.20202         0.606061                   100     100
    10       0.39899                     0.991469           2       2                  1                0.998197    1                           0.999503            0.191919        0.79798                    100     100
    11       0.5                         0.186367           1.7     1.93939            0.85             0.75707     0.969697                    0.950527            0.171717        0.969697                   70      93.9394
    12       0.60101                     0.0459512          0.2     1.64706            0.1              0.0914515   0.823529                    0.806144            0.020202        0.989899                   -80     64.7059
    13       0.69697                     0.0119742          0       1.42029            0                0.0219188   0.710145                    0.698171            0               0.989899                   -100    42.029
    14       0.79798                     0.00592631         0       1.24051            0                0.00824546  0.620253                    0.610839            0               0.989899                   -100    24.0506
    15       0.89899                     0.00339664         0.1     1.11236            0.05             0.004385    0.55618                     0.542698            0.010101        1                          -90     11.236
    16       1                           0.00187222         0       1                  0                0.00268928  0.5                         0.488152            0               1                          -100    0


Scoring History: 


timestamp            duration    iterations    negative_log_likelihood    objective
--  -------------------  ----------  ------------  -------------------------  -----------
    2019-05-15 11:00:47  0.000 sec   0             555.904                    0.693147
    2019-05-15 11:00:47  0.047 sec   1             178.636                    0.222738
    2019-05-15 11:00:47  0.047 sec   2             112.193                    0.139892
    2019-05-15 11:00:47  0.062 sec   3             85.9122                    0.107122
    2019-05-15 11:00:47  0.062 sec   4             75.5152                    0.0941587
    2019-05-15 11:00:47  0.062 sec   5             73.057                     0.0910935
    2019-05-15 11:00:47  0.062 sec   6             72.8665                    0.090856
    2019-05-15 11:00:47  0.078 sec   7             72.8649                    0.090854
    2019-05-15 11:00:47  0.078 sec   8             72.8649                    0.090854

In [10]:
# Coefficients that can be applied to the non-standardized data.
print(glm_model.coef())

{'Intercept': 27.87692087078464, 'Daily Time Spent on Site': -0.1966950113270224, 'Age': 0.14699298814289136, 'Area Income': -0.00013020677570598898, 'Daily Internet Usage': -0.06268513367372519, 'Male': -0.5392970117985854}


In [11]:
# Coefficients fitted on the standardized data (requires standardize = True, which is on by default)
print(glm_model.coef_norm())

{'Intercept': 1.6886784338118903, 'Daily Time Spent on Site': -3.1568417772378043, 'Age': 1.2706869405802226, 'Area Income': -1.7506282774499542, 'Daily Internet Usage': -2.710143301946955, 'Male': -0.2694465303066473}


In [12]:
# Print the Coefficients table
glm_model._model_json['output']['coefficients_table']

Coefficients: glm coefficients



names                     coefficients    std_error    z_value    p_value      standardized_coefficients
------------------------  --------------  -----------  ---------  -----------  ---------------------------
Intercept                 27.8769         3.07223      9.07383    0            1.68868
Daily Time Spent on Site  -0.196695       0.0227155    -8.65907   0            -3.15684
Age                       0.146993        0.0272385    5.39652    6.79451e-08  1.27069
Area Income               -0.000130207    2.01234e-05  -6.47041   9.77393e-11  -1.75063
Daily Internet Usage      -0.0626851      0.00765778   -8.18581   2.22045e-16  -2.71014
Male                      -0.539297       0.455624     -1.18364   0.236554     -0.269447

In [13]:
# Print the Standard error
print(glm_model._model_json['output']['coefficients_table']['std_error'])


[3.0722314292465556, 0.02271549899985897, 0.02723846770856031, 2.0123428634649435e-05, 0.007657780798422204, 0.45562413185764894]


In [14]:
predictions = glm_model.predict(test)
                               
predictions.head(5)

glm prediction progress: |████████████████████████████████████████████████| 100%


predict p0 p1 StdErr 
 
 
 0 0.996067 0.00393345 0.612624 
 0 0.90338 0.0966205 0.478406 
 1 1.30426e-05 0.999987 1.26424 
 0 0.964354 0.0356458 0.506453 
 1 3.03995e-05 0.99997 1.1998

In [15]:
## Get the AUC on the validation set
perf = glm_model.model_performance(test)
print (perf)


ModelMetricsBinomialGLM: glm
** Reported on test data. **

MSE: 0.02369804444066122
RMSE: 0.1539416916909166
LogLoss: 0.09510143545409801
Null degrees of freedom: 197
Residual degrees of freedom: 192
Null deviance: 274.48628350173834
Residual deviance: 37.66016843982282
AIC: 49.66016843982282
AUC: 0.9907152331394756
pr_auc: 0.9832395618791827
Gini: 0.9814304662789513
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.20978118730958129: 


0    1    Error    Rate
-----  ---  ---  -------  -----------
0      97   2    0.0202   (2.0/99.0)
1      3    96   0.0303   (3.0/99.0)
Total  100  98   0.0253   (5.0/198.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.209781     0.974619  97
max f2                       0.209781     0.97166   97
max f0point5                 0.688854     0.989474  93
max accuracy                 0.688854     0.974747  93
max precision                0.999998     1         0
max recall                   0.00588656   1         158
max specificity              0.999998     1         0
max absolute_mcc             0.688854     0.950708  93
max min_per_class_accuracy   0.209781     0.969697  97
max mean_per_class_accuracy  0.688854     0.974747  93

Gains/Lift Table: Avg response rate: 50.00 %, avg score: 78.48 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    score     cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------
    1        0.010101                    1.45121            2         2                  1                1.47678   1                           1.47678             0.020202        0.020202                   100       100
    2        0.020202                    1.3899             2         2                  1                1.44413   1                           1.46045             0.020202        0.040404                   100       100
    3        0.030303                    1.35123            2         2                  1                1.37593   1                           1.43228             0.020202        0.0606061                  100       100
    4        0.040404                    1.3196             2         2                  1                1.34024   1                           1.40927             0.020202        0.0808081                  100       100
    5        0.0505051                   1.30396            2         2                  1                1.31577   1                           1.39057             0.020202        0.10101                    100       100
    6        0.10101                     1.21214            2         2                  1                1.25747   1                           1.32402             0.10101         0.20202                    100       100
    7        0.151515                    1.1724             2         2                  1                1.1924    1                           1.28015             0.10101         0.30303                    100       100
    8        0.20202                     1.10906            2         2                  1                1.14129   1                           1.24543             0.10101         0.40404                    100       100
    9        0.30303                     0.965463           2         2                  1                1.03825   1                           1.17637             0.20202         0.606061                   100       100
    10       0.39899                     0.746376           2         2                  1                0.885808  1                           1.10649             0.191919        0.79798                    100       100
    11       0.5                         0.631459           0.8       1.75758            0.4              0.660724  0.878788                    1.01644             0.0808081       0.878788                   -20       75.7576
    12       0.60101                     0.605807           0.3       1.51261            0.15             0.619679  0.756303                    0.949755            0.030303        0.909091                   -70       51.2605
    13       0.69697                     0.573138           0.210526  1.33333            0.105263         0.592201  0.666667                    0.900526            0.020202        0.929293                   -78.9474  33.3333
    14       0.79798                     0.544041           0.4       1.21519            0.2              0.559182  0.607595                    0.857318            0.040404        0.969697                   -60       21.519
    15       0.89899                     0.507186           0.2       1.10112            0.1              0.522554  0.550562                    0.819704            0.020202        0.989899                   -80       10.1124
    16       1                           0.439314           0.1       1                  0.05             0.473958  0.5                         0.78478             0.010101        1                     

In [19]:
# Get the AUC on the training and testing data:
train_auc = glm_model.auc(train=True)
train_auc

test_auc = glm_model.auc(validation=True)
test_auc


TypeError: auc() got an unexpected keyword argument 'validation'

# Interpretation

In general, our model performed quite well F1 value of 0.97, precision value of 1.0, recall vale of 1.0 and AUC of 0.99